src/Native/Unix/System.Native/pal_random.c

```diff
@@ -2,109 +2,269 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#include <stdlib.h>
-#include <stdint.h>
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <sched.h>
 #include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
 #include <sys/stat.h>
-#include <fcntl.h>
-#include <assert.h>
-#include <unistd.h>
 #include <time.h>
-#include <errno.h>
+#include <unistd.h>
 
 #include "pal_config.h"
 #include "pal_random.h"
+
+#if defined(__linux__)
+#include <sys/syscall.h>
+#if defined(SYS_getrandom)
+# define PAL_GRND_NONBLOCK 1
+#endif // SYS_getrandom
+#endif // __linux__
+
 /*
 
 Generate random bytes. The generated bytes are not cryptographically strong.
 
 */
 
-void SystemNative_GetNonCryptographicallySecureRandomBytes(uint8_t* buffer, int32_t bufferLength)
+#if !HAVE_ARC4RANDOM_BUF
+static int GetUrandomFileDescriptor(void)
 {
-    assert(buffer != NULL);
-
-#if HAVE_ARC4RANDOM_BUF
-    arc4random_buf(buffer, (size_t)bufferLength);
-#else
-    static volatile int rand_des = -1;
-    long num = 0;
-    static bool sMissingDevURandom;
-    static bool sInitializedMRand;
+    static int fd = -1;
+    int newFd;
 
-    if (!sMissingDevURandom)
+    if (fd >= 0)
     {
-        if (rand_des == -1)
-        {
-            int fd;
+        return fd;
+    }
```

Comment:
There are no issues with the thread-safety around the reads of this field? e.g. is there any situation where the compiler/hardware might introduce a read such that it effectively became `int a = fd; if (fd >= 0) return a;`?